src/generations.rs

```diff
@@ -0,0 +1,143 @@
+use chrono::{DateTime, Local, TimeZone, Utc};
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::process;
+
+#[derive(Debug)]
+pub struct GenerationInfo {
+    pub generation: String,
+    pub date: String,
+    pub nixos_version: String,
+    pub kernel_version: String,
+    pub configuration_revision: String,
+    pub specialisations: Vec<String>,
+    pub current: bool,
+}
+
+pub fn generation_from_dir(generation_dir: &Path) -> Option<String> {
+    generation_dir
+        .file_name()
+        .and_then(|os_str| os_str.to_str())
+        .map(|generation_base| {
+            let no_link_gen = generation_base.trim_end_matches("-link");
+            no_link_gen
+                .rsplit_once('-')
+                .map(|(_, gen)| gen.to_string())
+                .unwrap_or_default()
+        })
+}
+
+pub fn describe_generation(generation_dir: &Path, current_profile: &Path) -> GenerationInfo {
+    let generation_number = generation_from_dir(generation_dir).unwrap_or_default();
+    let nixos_version = fs::read_to_string(generation_dir.join("nixos-version"))
+        .unwrap_or_else(|_| "Unknown".to_string());
+    let kernel_dir = generation_dir
+        .join("kernel")
+        .canonicalize()
+        .ok()
+        .and_then(|path| path.parent().map(|p| p.to_path_buf()))
+        .unwrap_or_else(|| PathBuf::from("Unknown"));
+
+    let kernel_version = fs::read_dir(kernel_dir.join("lib/modules"))
+        .and_then(|entries| {
+            let mut versions = vec![];
+            for entry in entries {
+                if let Ok(entry) = entry {
+                    if let Some(name) = entry.file_name().to_str() {
+                        versions.push(name.to_string());
+                    }
+                }
+            }
+            Ok(versions.join(", "))
+        })
+        .unwrap_or_else(|_| "Unknown".to_string());
```

Comment:
For darwin support you would have to use the `sw_vers -productVersion` or `/usr/libexec/PlistBuddy -c "Print:ProductVersion" /System/Library/CoreServices/SystemVersion.plist` to get the version.

Reply:
nh os is a nixos command, which is the scope of this PR

Reply:
In future if people wanted to support for it, I am giving little tips here to trace back to.